### PR TITLE
Avoid closure in IConnectionBuilder.Use

### LINKED ISSUE #46533: Avoid closure in IConnectionBuilder.Use
## Background and Motivation

Followup from https://github.com/dotnet/aspnetcore/issues/31463. The following code requires a closure to flow the context:

https://github.com/dotnet/aspnetcore/blob/f58f9546be23e6480569b2143e7aec168f666b2a/src/Servers/Connections.Abstractions/src/ConnectionBuilderExtensions.cs#L36-L46

While this is per connection rather than per request, it's still best to avoid it.

## Proposed API

```diff
namespace Microsoft.AspNetCore.Connections;

public static class ConnectionBuilderExtensions
{
     public static IConnectionBuilder Use(this IConnectionBuilder connectionBuilder, Func<ConnectionContext, Func<Task>, Task> middleware)
+    public static IConnectionBuilder Use(this IConnectionBuilder connectionBuilder, Func<ConnectionContext, Func<ConnectionContext, Task>, Task> middleware)
}
```

## Usage Examples

Before:
```c#
listenOptions.Use(async (context, next) =>
{
    await next();
});
```
New:
```C#
listenOptions.Use(async (context, next) =>
{
    await next(context);
});
```

## Alternative Designs

N/A

## Risks

Confusion about which overload to use. This can be mitigated via analyzers.

### PR DESCRIPTION
**Pull request by divyeshio**
# Avoid closure in IConnectionBuilder.Use

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Add new Use connectionbuilder extension method

## Description

Add a new IConnectionBuilder.Use overload that requires users to pass the ConnectionContext to the next function which will save allocations over the previous extension method.

Fixes #46533

### CODE REVIEW

**Comment by divyeshio** [comment]
> @divyeshio please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @dotnet-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @dotnet-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @dotnet-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@dotnet-policy-service agree

**Comment by JamesNK** [comment]
Could this change create ambiguous method errors when building?

These two `Use` overloads take a delegate with two arguments:
```cs
public static IConnectionBuilder Use(
    this IConnectionBuilder connectionBuilder,
    Func<ConnectionContext, Func<Task>, Task> middleware)
```
And
```cs
public static IConnectionBuilder Use(
    this IConnectionBuilder connectionBuilder,
    Func<ConnectionContext, ConnectionDelegate, Task> middleware)
```
It looks like implicit lambda args won't work anymore:
```cs
builder.Use((context, next) =>
{
    return Task.CompletedTask;
});
```
(I haven't tested this)

**Comment by divyeshio** [comment]
Similar changes for middleware have been already merged. This has been discussed at #31463

**Comment by JamesNK** [comment]
Ok! Are there any places the existing overload is used in aspnetcore that can be changed to the new method?

**Comment by divyeshio** [comment]
I tried to look for any references for previous method but couldn't find in anywhere including samples, tests.

Note: This is my first ever contribution, I might be missing something.

File: src/Servers/Connections.Abstractions/src/ConnectionBuilderExtensions.cs [hunk 1]
```diff
@@ -45,6 +45,27 @@ public static IConnectionBuilder Use(this IConnectionBuilder connectionBuilder,
         });
     }
 
+    /// <summary>
+    /// Add the given <paramref name="middleware"/> to the connection.
+    /// If you aren't calling the next function, use <see cref="Run(IConnectionBuilder, Func{ConnectionContext, Task})"/> instead.
+    /// <para>
+    /// Prefer using <see cref="Use(IConnectionBuilder, Func{ConnectionContext, ConnectionDelegate, Task})"/> for better performance as shown below:
```
Comment (amcasey) [inline src/Servers/Connections.Abstractions/src/ConnectionBuilderExtensions.cs:52]: Would it be more helpful to put this suggestion on the non-preferred overload?  I think that's what #31784 did.

**Comment by amcasey** [comment]
> Ok! Are there any places the existing overload is used in aspnetcore that can be changed to the new method?

Maybe [here](https://github.com/dotnet/aspnetcore/blob/3bdb2c15978749c61201a11398844507531bee6b/src/Servers/Kestrel/samples/Http2SampleApp/Program.cs#L53)?

**Comment by amcasey** [comment]
This does look pretty analogous to https://github.com/dotnet/aspnetcore/pull/31784, so @BrennanConroy might have thoughts.

**Comment by Tratcher** [comment]
There should at least be a test you can use this in.
https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/test/InMemory.FunctionalTests/ConnectionMiddlewareTests.cs

**Comment by divyeshio** [comment]
> There should at least be a test you can use this in. https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/test/InMemory.FunctionalTests/ConnectionMiddlewareTests.cs

None of the tests use this extension method

**Comment by Tratcher** [comment]
> > There should at least be a test you can use this in. https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/test/InMemory.FunctionalTests/ConnectionMiddlewareTests.cs
> 
> None of the tests use this extension method

I meant that you can copy an existing test and adapt it to use these extensions.

**Comment by JamesNK** [comment]
We want a test to use the new method to ensure the it works correctly.

**Comment by amcasey** [comment]
Seems sensible to me.  Good to go @Tratcher @JamesNK ?

**Review by Tratcher — APPROVED** [review]

**Pull request merged**